pkg/proxy/proxy.go

```diff
@@ -155,32 +150,138 @@ func (p *Proxy) SetIPContext(conn *networkservice.Connection, interfaceType netw
+		dstPrefix, err := p.IpamClient.Allocate(ctx, child)
 		if err != nil {
 			return err
 		}
 		dstIpAddrs = append(dstIpAddrs, dstPrefix.ToString())
 	}
 
-	conn.GetContext().IpContext = &networkservice.IPContext{}
 	if interfaceType == networking.NSE {
-		conn.GetContext().IpContext.SrcIpAddrs = srcIPAddrs
-		conn.GetContext().IpContext.DstIpAddrs = dstIpAddrs
-		conn.GetContext().GetIpContext().ExtraPrefixes = p.bridge.GetLocalPrefixes()
+		p.setNSEIpContext(conn.GetContext().GetIpContext(), srcIPAddrs, dstIpAddrs)
 	} else if interfaceType == networking.NSC {
-		conn.GetContext().IpContext.SrcIpAddrs = dstIpAddrs
-		conn.GetContext().IpContext.DstIpAddrs = srcIPAddrs
+		conn.GetContext().GetIpContext().SrcIpAddrs = dstIpAddrs
+		conn.GetContext().GetIpContext().DstIpAddrs = srcIPAddrs
 	}
 	return nil
 }
 
-func (p *Proxy) UnsetIPContext(conn *networkservice.Connection, interfaceType networking.InterfaceType) error {
+func (p *Proxy) setNSEIpContext(ipContext *networkservice.IPContext, srcIPAddrs []string, dstIpAddrs []string) {
+	if len(ipContext.SrcIpAddrs) == 0 && len(ipContext.DstIpAddrs) == 0 { // First request
+		ipContext.SrcIpAddrs = srcIPAddrs
+		ipContext.DstIpAddrs = dstIpAddrs
+		ipContext.ExtraPrefixes = p.Bridge.GetLocalPrefixes()
+		return
+	}
+	// The request is an update
+	if contains(ipContext.ExtraPrefixes, p.Bridge.GetLocalPrefixes()) &&
+		contains(ipContext.SrcIpAddrs, srcIPAddrs) &&
+		contains(ipContext.DstIpAddrs, dstIpAddrs) {
+		return
+	}
+	// remove old IPs, add new ones, and set the gateways
+	oldGateways := ipContext.GetExtraPrefixes()
+	ipContext.ExtraPrefixes = p.Bridge.GetLocalPrefixes()
+	ipContext.SrcIpAddrs = removeOldIPs(ipContext.SrcIpAddrs, oldGateways)
+	ipContext.DstIpAddrs = removeOldIPs(ipContext.DstIpAddrs, oldGateways)
+	ipContext.SrcIpAddrs = append(ipContext.SrcIpAddrs, srcIPAddrs...)
+	ipContext.DstIpAddrs = append(ipContext.DstIpAddrs, dstIpAddrs...)
+	// Find IPv4 gateway and IPv6 Gateway
+	gatewayV4 := ""
+	gatewayV6 := ""
+	for _, gw := range ipContext.ExtraPrefixes {
+		ip, _, err := net.ParseCIDR(gw)
+		if err != nil {
+			continue
+		}
+		if ip.To4() != nil {
+			gatewayV4 = ip.String()
+			continue
+		}
+		gatewayV6 = ip.String()
+	}
+	// Replace the nexthops in the policy routes with the up to date gateways
+	for _, policyRoute := range ipContext.GetPolicies() {
+		ipv4 := true
+		ip, _, err := net.ParseCIDR(policyRoute.From)
+		if err != nil {
+			continue
+		}
+		if ip.To4() == nil {
+			ipv4 = false
+		}
+		for _, route := range policyRoute.Routes {
+			if ipv4 {
+				route.NextHop = gatewayV4
+			} else {
+				route.NextHop = gatewayV6
+			}
+		}
+	}
+}
+
+// removes all IPs in the ips list that are in the same subnet as any of the gateway
+func removeOldIPs(ips []string, gateways []string) []string {
+	gws := []*net.IPNet{}
+	for _, ip := range gateways {
+		_, n, err := net.ParseCIDR(ip)
+		if err != nil {
+			continue
+		}
+		gws = append(gws, n)
+	}
+	ipsMap := listToMap(ips)
```

Comment:
What's the point of this conversion considering the rest of the function?

Reply:
true, I will solve it